http/client.go

```diff
@@ -8,114 +8,165 @@ import (
+		files, err := ioutil.ReadDir(CABundleDir)
+		if err != nil {
+			return err
+		}
+
+		for _, f := range files {
+			if strings.Contains(f.Name(), ".pem") {
+				caCertFilePath := filepath.Join(CABundleDir, f.Name())
+				if err := addCert(caCertFilePath, certPool); err != nil {
+					return err
 				}
-				_, err := certs[0].Verify(opts)
-				return err
 			}
-
 		}
+		return nil
 	}
-
-	return &http.Client{
-		Timeout:   httpTimeout,
-		Transport: t,
-	}, nil
 }
 
-func getCertPool(certFile string, certDirectory string) (*x509.CertPool, error) {
-	caCertPool := x509.NewCertPool()
+// WithAcceptInvalidHostname allows the client to call the acceptInvalidHostname host
+// instead of the host from the certificates.
+func WithAcceptInvalidHostname(acceptInvalidHostname string) ClientOption {
+	return func(c *http.Client) error {
+		if acceptInvalidHostname == "" {
+			return nil
+		}
 
-	if certFile != "" {
-		if err := addCert(filepath.Join(certDirectory, certFile), caCertPool); err != nil {
-			if os.IsNotExist(err) {
-				if err = addCert(certFile, caCertPool); err != nil {
-					return nil, err
+		transport, ok := c.Transport.(*http.Transport)
+		if !ok {
+			return ErrInvalidTransportType
+		}
+		// Default validation is replaced with VerifyPeerCertificate.
+		// Note that when InsecureSkipVerify and VerifyPeerCertificate are in use,
+		// ConnectionState.VerifiedChains will be nil.
+		transport.TLSClientConfig.InsecureSkipVerify = true
+		// While packages like net/http will implicitly set ServerName, the
+		// VerifyPeerCertificate callback can't access that value, so it has to be set
+		// explicitly here or in VerifyPeerCertificate on the client side. If in
+		// an http.Transport DialTLS callback, this can be obtained by passing
+		// the addr argument to net.SplitHostPort.
+		transport.TLSClientConfig.ServerName = acceptInvalidHostname
+		// Approximately equivalent to what crypto/tls does normally:
+		// https://github.com/golang/go/commit/29cfb4d3c3a97b6f426d1b899234da905be699aa
+		transport.TLSClientConfig.VerifyPeerCertificate = func(certificates [][]byte, _ [][]*x509.Certificate) error {
+			certs := make([]*x509.Certificate, len(certificates))
+			for i, asn1Data := range certificates {
+				cert, err := x509.ParseCertificate(asn1Data)
+				if err != nil {
+					return errors.New("tls: failed to parse certificate from server: " + err.Error())
 				}
+				certs[i] = cert
+			}
+
+			opts := x509.VerifyOptions{
+				Roots:         transport.TLSClientConfig.RootCAs, // On the server side, use config.ClientCAs.
+				DNSName:       acceptInvalidHostname,
+				Intermediates: x509.NewCertPool(),
+				// On the server side, set KeyUsages to ExtKeyUsageClientAuth. The
+				// default value is appropriate for clients side verification.
+				// KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
 			}
+			for _, cert := range certs[1:] {
+				opts.Intermediates.AddCert(cert)
+			}
+			_, err := certs[0].Verify(opts)
```

Comment:
I know this wasn't touched, but shouldn't Verify be run for each cert?

Reply:
As far as I can tell this should be fine, as per the example here: https://github.com/golang/go/commit/29cfb4d3c3a97b6f426d1b899234da905be699aa

The first cert is the one we want to verify, and the rest are the issue chain, which get added into `opts`.